### PR TITLE
Modify rar-service design doc

### PR DESCRIPTION
**Pull request by rokonec**
### Context
During implementing of RAR code we realize we need to revisit some of our decisions from https://github.com/dotnet/msbuild/issues/5536

### Changes Made
Updating .md document to reflect our current plan.

### CODE REVIEW

File: documentation/specs/rar-as-service.md [hunk 1]
```diff
@@ -42,14 +42,35 @@ This step will create new process which will act as RAR node. It will also pass
...
 
+RAR Node will adopt existing MSBuild infrastructure code, NodeProviderOutOfProcTaskHost and related, used for invoking tasks out of process.
+
+This code already solved many aspect of 'Out of process task invocation':
+- logging
+- environmental variables
+- current directory path
+- current culture
+- cancellation
+- etc...
```
Comment (ladipro) [inline documentation/specs/rar-as-service.md:53]: And, most importantly, serializing parameters passed to and from the out-of-proc task.

File: documentation/specs/rar-as-service.md [hunk 2]
```diff
@@ -42,14 +42,35 @@ This step will create new process which will act as RAR node. It will also pass
...
 
 Execution should be the same as it is now.
 
 There is already some layer of separation between Task interface and actual execution method. We will leverage this, and put the decision logic if to run locally or not into the "wrapper" method and so we will not have to modify this and in server-side execution we will directly call the internal execution method.
 
+#### RAR Concurrency
+
 There is one big concern and that is how to handle multiple requests at once. As right now, RAR task is not prepared for multi-thread use.
 
+One of the biggest challenges with RAR as service, is to make execution and caching of RAR task thread-safe, since in most cases there will be multiple clients request data from it at once.
```
Comment (ladipro) [inline documentation/specs/rar-as-service.md:65]: nit: typo "... clients request**ing** data ..." ?

File: documentation/specs/rar-as-service.md [hunk 3]
```diff
@@ -42,14 +42,35 @@ This step will create new process which will act as RAR node. It will also pass
...
+
+Following areas has to be addressed to allow concurrent execution of RAR:
```
Comment (ladipro) [inline documentation/specs/rar-as-service.md:67]: nit: typo "... ha**ve** ..."

File: documentation/specs/rar-as-service.md [hunk 4]
```diff
@@ -42,14 +42,35 @@ This step will create new process which will act as RAR node. It will also pass
...
+
+- thread safety (static variables, shared data structures, caching, ...)
+- environmental variables virtualization
+- current directory virtualization
+- current culture isolation
```
Comment (ladipro) [inline documentation/specs/rar-as-service.md:72]: ❤️ 
How confident are you that this list is complete?

File: documentation/specs/rar-as-service.md [hunk 5]
```diff
@@ -42,14 +42,35 @@ This step will create new process which will act as RAR node. It will also pass
 
 We will use Mutex (as in [Roslyn](https://github.com/dotnet/roslyn/blob/838874b1b817db84ce146bef690cc95a39c213a5/src/Compilers/Server/VBCSCompiler/BuildServerController.cs#L143)) to ensure we don't create two RAR nodes at once. Its name must encode whether it is the user's only RAR node, including user name, administrator privileges, and some initial settings for the node. Such a name could be: `MSBuild.RAR.ostorc.7`, where **MSBuild.RAR** is its prefix, **ostorc** is the user who called MSBuild, and **7** represents encoded settings (flag enum).
```
Comment (ladipro) [inline documentation/specs/rar-as-service.md:43]: Do we still need a mutex? If multiple nodes try to create a pipe of the same name, all but one will fail. Isn't it enough to implement mutual exclusion?

**Review by ladipro — COMMENTED** [review]
Can't comment on this part in-line as it's not being changed in the PR:

_Connecting to the RAR node will not require any discovery of processes on computer._

Do you think this is still true? Or are we reusing the existing handshake mechanism?

File: documentation/specs/rar-as-service.md (hunk 4 above)
Comment (rokonec) [reply documentation/specs/rar-as-service.md:72]: adding etc.... :-)

File: documentation/specs/rar-as-service.md (hunk 5 above)
Comment (rokonec) [reply documentation/specs/rar-as-service.md:43]: It should be enough -AFAIK -  I wasn't focusing at that area. Will update it.

Comment (rokonec) [reply documentation/specs/rar-as-service.md:43]: @Ladi I an unaware of 'user's only RAR node' means. Can you please elaborate?

**Comment by rokonec** [comment]
> Can't comment on this part in-line as it's not being changed in the PR:
> 
> _Connecting to the RAR node will not require any discovery of processes on computer._
> 
> Do you think this is still true? Or are we reusing the existing handshake mechanism?

We plan to reuse existing handshaking. Connection to RAR service will be done by opening named-pipe including existing handshaking.

File: documentation/specs/rar-as-service.md (hunk 5 above)
Comment (ladipro) [reply documentation/specs/rar-as-service.md:43]: This refers to the requirement that there shouldn't be more than one RAR node running for the given user (in a given session?). In other words, we want a singleton process, same as the Roslyn compilation server, for example.

(Looks like you tagged somebody else in your comment.)

**Review by ladipro — APPROVED** [review]

**Pull request merged**